cli/CampaignConfigValidation/CampaignValidator.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation;
+
+use WMDE\Fundraising\Frontend\BucketTesting\Campaign;
+use WMDE\Fundraising\Frontend\BucketTesting\CampaignCollection;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\CampaignValidationRuleInterface;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\DefaultBucketRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\StartAndEndTimeRule;
+use WMDE\Fundraising\Frontend\Cli\CampaignConfigValidation\Rule\UniqueBucketRule;
+
+/**
+ * Validates campaign data by looking for inconsistencies and logical errors
+ * YAML validity is checked in during loading of campaign configuration
+ *
+ * @see \WMDE\Fundraising\Frontend\BucketTesting\CampaignConfiguration
+ * @license GNU GPL v2+
+ */
+class CampaignValidator {
+
+	private $campaignCollection;
+	private $hasValidated = false;
+	private $errorLogger;
+	private $rules = [];
+
+	public function __construct( CampaignCollection $campaignCollection, ValidationErrorLogger $errorLogger ) {
+		$this->campaignCollection = $campaignCollection;
+		$this->errorLogger = $errorLogger;
+		$this->rules = [ new DefaultBucketRule(), new StartAndEndTimeRule(), new UniqueBucketRule() ];
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errorLogger->getErrors();
+	}
+
+	private function validate() {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		foreach ( $this->campaignCollection->getIterator() as $campaign ) {
```

Comment:
The `IteratorAggregate` interfaces ensures that you can iterate `CampaignCollection` without calling `getIterator`, the will be used by PHP internally.